### DISCUSSION TITLE
Add multiple error returning in JSON Serializing

### DISCUSSION THREAD

**Discussion by Vinicius-Evangelista** [dotnet/runtime#109372] [Ideas]
Hi everyone,

This is my first discussion in my career, and I’m excited to contribute an idea.

While working with JSON serialization, I encountered a scenario where the JSON had multiple errors. Currently, the serializer identifies one error at a time, which requires multiple iterations to find and fix all issues. I believe it would be beneficial if the serializer could return all errors in one go, which could greatly improve the debugging process.

This change would save time and make it easier for developers to resolve JSON issues quickly. I'd love to hear your thoughts on this idea!

Here is an example to simplify what I'am trying to say:

```csharp
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

public class JsonValidator
{
    public List<JsonError> ValidateJson(string jsonString)
    {
        var errors = new List<JsonError>();
        
        // Define JSON options with custom error handling
        var options = new JsonSerializerOptions
        {
            AllowTrailingCommas = false, // Disallow trailing commas
            PropertyNameCaseInsensitive = true // Make property names case-insensitive
        };

        try
        {
            // Try to deserialize the JSON to a specific class, catching errors
            JsonSerializer.Deserialize<MyClass>(jsonString, options);
        }
        catch (JsonException ex)
        {
            // Parse each error in the JSON and add it to the error list
            errors.Add(new JsonError
            {
                Path = ex.Path ?? "Unknown path",
                Message = ex.Message
            });
        }

        return errors;
    }
}

// A sample class to represent your JSON structure
public class MyClass
{
    public string Name { get; set; }
    public int Age { get; set; }
    public string Email { get; set; }
    public bool IsActive { get; set; }
}

// Class to represent each individual JSON error
public class JsonError
{
    public string Path { get; set; } // Path to the JSON property with an error
    public string Message { get; set; } // Error message
}

// Example of how to use the JsonValidator
public class Program
{
    public static void Main()
    {
        string json = @"{
            ""Name"": ""Vinicius"",
            ""Age"": ""Twenty"",
            ""Email"": ""vinicius@example"",
            ""IsActive"": tru
        }";

        var validator = new JsonValidator();
        List<JsonError> errors = validator.ValidateJson(json);

        // Display all errors found in the JSON
        foreach (var error in errors)
        {
            Console.WriteLine($"Path: {error.Path}, Message: {error.Message}");
        }
    }
}
```

**Comment by spydacarnage** [comment]
My immediate thought on why it works as it currently does is that, especially for formatting errors, such as missing commas or other unexpected tokens, once a single error is found, _everything else_ is an error, because the reader no longer knows what to expect next.

**Reply by Vinicius-Evangelista** [reply]
Yes, I strongly agree.
But sometime I faced some errors happens when some required field is missing in the json that is required in the target class.
If all this errors could return it would awesome.

**Reply by spydacarnage** [reply]
But then you find yourself in a scenario when sometimes you get all the errors and sometimes you don't.
And if you only get one error, you don't know if that's all of them, or just that it can only return one this time.

**Reply by Vinicius-Evangelista** [reply]
Yes.
What if you could explicit pass a flag to when we want all the errors or not?

**Reply by spydacarnage** [reply]
It doesn't matter whether you _want_ them all - if it can't provide them all, you can't know if you have them all or not.

**Comment by lakeman** [comment]
JSON syntax errors should fail fast. On the first unexpected or unhandled token.

Semantic errors during serialisation? I can imagine how you could continue after the first error and fail at the end. But that's a significant change to the serialisation interface. Including all custom `JsonConverter` implementations, where currently the only way to indicate a failure is to throw an exception. Particularly because parsing the json and serialising the C# result occur in one pass, for performance reasons.
